### ISSUE TITLE
Pretty good results of val_iou_score but very bad in segmentation of validation data

### ISSUE THREAD

**Issue by mirasmd** [qubvel/segmentation_models#207]
Hi, first of all thanks for this very nice and helpful repo! 

I am trying out a few combinations of models and backbones at the moment, and i've got a problem, that i wasn't able to solve so far. 

While training the models i get very good results for iou_score, the val_iou_score (>0.9) and also the loss plots look as they should (smooth and converge to very small values). 
But when i use some of the trained models on the validation dataset, there is nearly no segmentation visible. The pictures are almost single-color with at most a few pixels beeing detected as the other class.. 

I have already tried different batch_sizes, losses (dice_loss, bce_jaccard_loss, mean_squared_error), optimizers (Adam, sgd) and metrics (iou_score, accuracy) and the problem remains the same... 

For all of the model-backbone combinations, the val_iou_scores are very high, but only some of them perform good at the segmenation. So i can't really predict the quality of a model based on the metrics. I'm quite new to image segmentation and keras so i might have missed something obvious, but maybe you can help me with this?

**Comment by JordanMakesMaps** [comment]
I had a similar problem, where all the metrics I used showed really good progress over a short period of time only to get predictions where all of the image was labelled one class 😒

What turned out to be my issue was:

- how I was organizing the masks (I have a 'background' pixels in each image that denote parts I want to ignore but takes up a lot of pixels, and I found that I was organizing the masks in such a way that the background pixels were representing an actual class. Therefore labelling each image entirely as background was the optimal choice for the algorithm).

- not using the `preprocess_input` function that comes with each backbone type

Hope that helps!

**Comment by mirasmd** [comment]
Thanks for your tips! 
I just deleted the usage of `preprocess_input` and I think the results improved a little bit. I will examine this a little bit further... 
But still, the calculated iou scores are far away from the real segmentation results.

What exactly do you mean by "labelling each image entirely as background"?
I might have the same problem here, because I have two classes, where one is only the background (path and not-path). 

But I'm quite sure that the training itself is running fine, because for some backbone-model combinations, I get very good segmentation results. I just need the scores to be able to compare the models qualities.

**Comment by JordanMakesMaps** [comment]
Sorry, I meant that _not_ using the `preprocess_intput` function gave me worse results, so now I use it. Not sure if that's how it was interpreted based on how I wrote it (corrected).

`What exactly do you mean by "labelling each image entirely as background"?`
  
I'm working with images where there are a lot of things going on in them, but some of the classes that exist in the images aren't important to me so I label them as background. But really I wanted them to be _ignored_. So I needed to adjust how I provide the one-hot-encoded masks during the training process so that they aren't interpreted as a class called 'background`. This sound different than your problem which sounds like a binary classification between path and not-path.

If you get good results using certain backbones and architecture combinations, but worse results with other combinations, it might just be that some combinations are crap for your particular dataset.
How imbalanced are the classes that you're working with? I found that some are more prone to overfitting to the majority class and labeling all pixels as that class; freezing the encoder helped me a lot with this.

Sorry that I can't answer the question definitely, though. But, if you wanted to add the combinations that work well for you here #205, we might be able to compile a list of combinations that generally do better than others.

**Comment by mirasmd** [comment]
Oh okay, now I got it! 

Yes, you're right, I'm doing a binary classification in my task and I think each class is about the half of all pixels. 

But the problem was not really that the results were bad, but that the metrics didn't match the actual results. But I'll just use other metrics then..
But thanks for all your tips! I think I'll then go for the combinations that work well. When I have some results I will share them :) 

I've still got one more question on the `preprocess_input` function:
I am using it in the training of the network, but do I also have to use it when I use the trained model? 
So do I have to preprocess the input before the segmentation? 
I think it might make sense, because I want the same kind of input in training and final segmentation but I'm not sure about it. 

Thanks in advance!

**Comment by JordanMakesMaps** [comment]
`So do I have to preprocess the input before the segmentation?`

Absofreakinlutely! Definitely use `preprocess_input` on images that you pass to the train model to be predicted. The reason being is that, if you use some form of standardization (which is what `preprocess_input` is doing) on your images during the training process, your model is expecting that format during testing as well. The `preprocess_input` function is different for each backbone (values between [-1, 1], or [0, 1] or [0, 255] as an example). So you if you train on one format of images, but then test on another format, then your results are going to be super wonky.

**Comment by mirasmd** [comment]
Oh, I didn't know that `preprocess_input` is different for each backbone, my fault! That helped a lot, thank you!
